driver/normalizer/strconv_test.go

```diff
@@ -0,0 +1,139 @@
+package normalizer
+
+import (
+	"regexp"
+	"testing"
+	"unicode/utf8"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const msg = "test case %d failed"
+
+var testCasesUnquote = []struct {
+	quoted   string
+	unquoted string
+	// isn't empty only for cases where quote(unqote(s)) != s
+	// e.g when we loose the informaiton wich original escape sequence was used
+	// Golang unquote() defaults to hex format, so it's used as canonical one.
+	canonicalQuoted string
+}{
+	{"'a'", "a", ""},
+	{"'\\x00'", "\u0000", ""},
+	{`'\0'`, "\u0000", "'\\x00'"},
+	{`'\0something\0'`, "\u0000something\u0000", "'\\x00something\\x00'"},
+	{`'\0something\0else'`, "\u0000something\u0000else", "'\\x00something\\x00else'"},
+	{"'\u0000123\\0s'", "\u0000123\u0000s", "'\\x00123\\x00s'"},
+}
+
+func TestUnquoteSingle(t *testing.T) {
+	for i, test := range testCasesUnquote {
+		s, err := unquoteSingle(test.quoted)
+		require.NoError(t, err, msg, i)
+
+		require.Equal(t, test.unquoted, s, msg, i)
+	}
+}
+
+func TestUnquoteSingleAndQuoteBack(t *testing.T) {
+	for i, test := range testCasesUnquote {
+		u, err := unquoteSingle(test.quoted)
+		require.NoError(t, err, msg, i)
+
+		q := quoteSingle(u)
+		if test.canonicalQuoted != "" {
+			assertEquals(t, test.canonicalQuoted, q, i)
+		} else {
+			assertEquals(t, test.quoted, q, i)
+		}
+	}
+}
+
+func assertEquals(t *testing.T, quoted, actual string, i int) {
+	success := assert.Equal(t, quoted, actual, msg, i)
+	if !success {
+		printDebug(t, quoted, actual)
+		t.FailNow()
+	}
+}
+
+func printDebug(t *testing.T, quoted, actual string) {
+	t.Logf("\texepcted: len=%d", len(quoted))
+	for _, c := range quoted {
+		t.Logf("%x - %#U", c, c)
+	}
+	t.Logf("\n\tactual: len=%d", len(actual))
+	for _, c := range actual {
+		t.Logf("%x - %#U", c, c)
+	}
+}
+
+func BenchmarkReplacingNullEscape_Iterative(b *testing.B) {
+	b.ReportAllocs()
+	for n := 0; n < b.N; n++ {
+		for _, test := range testCasesUnquote {
+			replaceEscapedMaybeIter(test.quoted, '0', '\x00')
+		}
+	}
+}
+
+func replaceEscapedMaybeIter(s string, old, new rune) string {
```

Comment:
I think we needn't keep the benchmarks for the previous implementation or the regexp experiment.

Reply:
It makes me sad, loosing information that way: this clearly communicates that multiple implementations were considered and tested, they also can be removed subsequently instead of maintenance, to avoid the burden.

This effectively makes it const nothing to keep those tests, if it's clearly marked as test-only.

Or am I missing something here?

Reply:
I understand where you're coming from, but I think it isn't a good idea to keep around experiments in the main code or tests for a package. that doesn't mean it's not useful or interesting, just that it's not relevant to anyone except us.

Fortunately, I don't think anything has to be lost—we could put them into a separate package in a subdirectory, so that someone running tests or benchmarks for this package doesn't pick up a bunch of experiments that aren't relevant to the current implementation. As someone who uses the tests for a package as a primary mechanism for understanding how it works, my own preference is to have a clear separation of "production" from "experiments".

For the reader, if we want to say we tried various things we could just as well put in a comment with the benchmark results: "We tested regexp replacement and manual expansion against the current implementation and found this was fastest".

So: I don't think it needs deleted entirely, but I do think it shouldn't live here.

Reply:
Thank you for explanation, @creachadair that makes sense.

I was missing the perspective of using package API surface as a communication tool for the state of public production VS internal experiments.

> Fortunately, I don't think anything has to be lost

indeed. And another way that may work for this case - is to have it in SVC history.

Updated in 9085daba243ab77d428c35a5a16d0a7d3a8dd148